src/erlcloud_waf.erl

```diff
@@ -0,0 +1,1188 @@
+
+configure(AccessKeyID, SecretAccessKey, Host) ->
+    put(aws_config, new(AccessKeyID, SecretAccessKey, Host)),
+    ok.
+
+-spec configure(string(), string(), string(), non_neg_integer()) -> ok.
+
+configure(AccessKeyID, SecretAccessKey, Host, Port) ->
+    put(aws_config, new(AccessKeyID, SecretAccessKey, Host, Port)),
+    ok.
+
+default_config() ->
+    erlcloud_aws:default_config().
+
+%%%------------------------------------------------------------------------------
+%%% AWS WAF API Functions
+%%%------------------------------------------------------------------------------
+
+%%%------------------------------------------------------------------------------
+%%% CreateByteMatchSet 
+%%
+%% http://docs.aws.amazon.com/waf/latest/APIReference/API_CreateByteMatchSet.html
+%%%------------------------------------------------------------------------------
+-spec create_byte_match_set/2 ::
+        (ChangeToken :: string() | binary(),
+         Name :: string() | binary()) ->
+        waf_return_val().
+create_byte_match_set(ChangeToken, Name) ->
+    create_byte_match_set(ChangeToken, Name, default_config()).
+
+-spec create_byte_match_set/3 ::
+        (ChangeToken :: string() | binary(),
+         Name :: string() | binary(),
+         Config :: aws_config()) ->
+        waf_return_val().
+create_byte_match_set(ChangeToken, Name, Config) ->
+    Json = [{<<"ChangeToken">>, to_binary(ChangeToken)},
+            {<<"Name">>, to_binary(Name)}],
+    waf_request(Config, "CreateByteMatchSet", Json).
+
+
+%%%------------------------------------------------------------------------------
+%%% CreateIPSet
+%%
+%% http://docs.aws.amazon.com/waf/latest/APIReference/API_CreateIPSet.html
+%%%------------------------------------------------------------------------------
+-spec create_ip_set/2 ::
+        (ChangeToken :: string() | binary(),
+         Name :: string() | binary()) ->
+        waf_return_val().
+create_ip_set(ChangeToken, Name) ->
+    create_ip_set(ChangeToken, Name, default_config()).
+
+-spec create_ip_set/3 ::
+        (ChangeToken :: string() | binary(),
+         Name :: string() | binary(),
+         Config :: aws_config()) ->
+        waf_return_val().
+create_ip_set(ChangeToken, Name, Config) ->
+    Json = [{<<"ChangeToken">>, to_binary(ChangeToken)},
+            {<<"Name">>, to_binary(Name)}],
+    waf_request(Config, "CreateIPSet", Json).
+
+
+%%%------------------------------------------------------------------------------
+%%% CreateRule
+%%
+%% http://docs.aws.amazon.com/waf/latest/APIReference/API_CreateRule.html
+%%%------------------------------------------------------------------------------
+-spec create_rule/3 ::
+        (ChangeToken :: string() | binary(),
+         MetricName :: string() | binary(),
+         Name :: string() | binary()) ->
+        waf_return_val().
+create_rule(ChangeToken, MetricName, Name) ->
+    create_rule(ChangeToken, MetricName, Name, default_config()).
```

Comment:
Since the form order in the [AWS Console](https://console.aws.amazon.com/waf/home?region=us-east-1#/rules/new) is `Name` then `CloudWatch metric name`, I believe it's enough to suggest the preferable order.